### ISSUE TITLE
Vector tile layers

### ISSUE THREAD

**Issue by charlenni** [Mapsui/Mapsui#1478]
Hello,

I want to ask if someone wants to have a look into my first version of [VectorTileLayers](https://github.com/charlenni/VectorTileLayers), an implementation for handling vector tiles with OpenMapTiles format in Mapsui. It is far from ready, but I like to know if it could work and if it is fast enough to be used on different systems. It has a WPF, Android and iOS sample to test it.

Would be nice to hear from you.

Thank you for your help.

**Comment by inforithmics** [comment]
I had a look at your implementation, It looks great. 

I made a Performance Evaluation and the most expensive function is DrawPath on a SKCanvas.  So rendering only once to a SKCanvas and than caching that would certainly improve Performance. But this is an ongoing discussion how to render and draw.

<img width="557" alt="image" src="https://user-images.githubusercontent.com/2290863/154798211-597e83fe-6241-4fba-ad2d-da551d51402b.png">

**Comment by charlenni** [comment]
Thank you for looking into this. Then I know, that it's working.

Caching is in this case difficult. The SKPaint is dependent from the zoom level and could change from draw cycle to draw cycle. I tried to optimize the SKPath drawing by creating one big path where possible. And I tried to use always the same SKPath, so that Skia could optimize the drawing. Other vector tile libraries upload the vertices to the OpenGL memory and so don't have to upload them again and again.

But beside this, I was impressed, how fast Skia could render the paths. I get a drawing time of around 25 ms against of 2.5 with raster tiles.

**Comment by inforithmics** [comment]
I looked at the Code more in Depth and saw that you are Rendering to SkObjets like SkPath and then on draw drawing these objects.

But wouldn't it be possible to Render the whole Tile to a SkPicture, and cache this

    ```
var pictureRecorder = new SKPictureRecorder();
    using var skCanvas = pictureRecorder.BeginRecording(new SKRect(0, 0, width, height));
    // to the Rendering for the Tile once here.
    var renderedSkPicture =pictureRecorder.EndRecording();
```

and then in the Draw only write once the SkPicture (Because SkPicture is a Vector Picture this can be zoomed without pixelation)

And should be faster because every DrawCall has a .NET Managed -> Native Translation going on. On Drawing the whole SkPicture this is only one call.

**Comment by charlenni** [comment]
No, that isn't possible, because the paint depends on the zoom (not only the zoom level as an integer, but as a float). So it would only be possible to use SKDrawable, but this needs even more converting, because the Draw function is called each time the SKDrawable is drawn.

What I'm now trying, is to reduce the calls to `DrawPath()`. With this, I don't call `DrawPath()` if the path isn't visible on screen. Skia checks this by itself, but I save the function call to the native library function.

The first optimization was to create one big SKPath for all lines that have the same style. But this is only possible for lines. Polygons could lie on top of each other. Skia would optimize this and delete the overlap. An example is the following. See the Roseraie Princess Grace. There are two identical polygons on top of each other with opacity.

![grafik](https://user-images.githubusercontent.com/4381223/154855068-84653ed8-4e00-4a30-b9eb-2cb040b0c5de.png)

So I have to draw each polygon by its own. Perhaps this could be improved by checking, if two polygons overlap and if not, add them. Draw as much SKPath with the same SKPaint as possible.

Also, the drawing of symbols could be improved. At the moment, I draw all symbols of a tile that fit into it, even if it isn't visible on the screen.
